[HSDS-269 HSDS-265 HSDS-268 HSDS-270] Various fixes

tinkertrain opened the pull request:
## Table focus state is missing left blue shadow in Safari
https://helpscout.atlassian.net/browse/HSDS-269

### Problem
Turns out some browsers (safari 😒) don't like to apply `box-shadow` to elements with a `display` property value other than `block` or `inline-block`. Since `<tr>` are `display: table-row`, the focus indicator was not shown.

### Solution
Instead of using an inset box shadow, we use a css gradient to achieve the same effect, this also fixes the little imperfection on the focus indicators (as shown on the ticket).

<img width="1136" alt="Screen Shot 2022-06-20 at 5 09 07 PM" src="https://user-images.githubusercontent.com/1414086/174634575-949419ce-6d0f-4abf-9929-f8782b1ca9e8.png">
<img width="792" alt="Screen Shot 2022-06-20 at 5 08 05 PM" src="https://user-images.githubusercontent.com/1414086/174634587-cac5f269-bc95-4794-b1d0-1b92eace0572.png">


## Input char validator badge is not hiding when reached limit and then cleared text
https://helpscout.atlassian.net/browse/HSDS-265

### Problem
The character validation state was incorrect in some instances, and the use of `Animate` was also causing some issues which caused some visual glitches on updateing the badge.


### Solution
We fix the state and remove the use of animations for the badge as their visual effect is negligible. Also updated with better tests for this feature.

![gif](https://cleanshot-cloud-fra.s3.eu-central-1.amazonaws.com/media/20196/UnZ7LMH4shkyNa6jK5HjgzZowC2cKNjmPbxc9gor.gif?X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEBYaDGV1LWNlbnRyYWwtMSJIMEYCIQDtd8hOTbP77K9JO%2FD0zSzYv%2FXKSL%2F0SNCUtk%2B41Dhj1QIhANtp1TyRHHgXhUavqNVeA44wmXcMXx0xEme2zQaHb0BSKqECCC8QABoMOTE5NTE0NDkxNjc0Igw37aWfmrKZ5rgcGSsq%2FgE0Q%2FHsLeBgFpNb%2FxwjG5xHR0qNalf7W2VovUKbuu4gwqR0gPt4Z50CDo6RVnAoLa8LxFB9GCXMDWJbMci4NEuMz1%2BHMjSVzPv5d9H0Ga3etWXpF4P4qkj8VaPlJV16EeOTGfR%2BKRBe0IKpsfVF5yQvZ9wqPGYS8nq%2FPFMvcd0RBcMb%2BJtgt9PWEIGwoWepp5hYisF8HaP0Xz%2FH65xSmCd4cgvRSlREutC5em9JQy5SIFN0u%2FRVJKlGlUQnu1WZzNOVWX4FiARfzeMTGsfWg%2BtPa6kDwE%2B6Dmu6FHTf2UDUN%2FqrtGAcOhtGRWHbgedvbV2bSDK7h0uq8JB9WIqxjjCF%2BsGVBjqZAVedkSq%2BsxnDfzVsk7ZzDbY2mUbLlSlZ9fKGx6PpGzq85TAjBrtbZa0%2FCilvKyTHnk%2Fc6MTZ4ZMmYB12pINBtnp4ljSDKGIb4%2FgBnM%2B%2BA%2F0fkk0frpWa%2B%2BRgto17IfjBq3dCXcgmQbIJPS9I%2BeoNz6MSqFBCNd2v40lb4hZtRGlQnUXg6Enc5KPyUaWaoysc9%2BDBH6watL9pIA%3D%3D&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIA5MF2VVMNLX55SSPG%2F20220620%2Feu-central-1%2Fs3%2Faws4_request&X-Amz-Date=20220620T152541Z&X-Amz-SignedHeaders=host&X-Amz-Expires=300&X-Amz-Signature=b177fa5088fb88fe80f9e16109cc6eabbdd3fe884025ac1088fcc70c85d29362)

## Nested SimpleModal instance not compatible with voice over (Safari)
https://helpscout.atlassian.net/browse/HSDS-268

### Problem
For some reason, safari wasn't updating the context when opening a modal on top of another modal using VoiceOver.

### Solution
Use `aria-hidden` to hide elements from the voice reader.

📹 https://i.hlp.sc/lbJRc0

## EditableField: validation icon is misaligned
https://helpscout.atlassian.net/browse/HSDS-270

## Problem
The validation icon is misaligned when the value of the input is larger than the width of the field

## Solution
We add some padding when the icon is present.

<img width="286" alt="Screen Shot 2022-06-20 at 5 00 53 PM" src="https://user-images.githubusercontent.com/1414086/174634852-4ddef43f-c694-4a64-9e48-a081e3013210.png">

## Other

We update the babel and storybook's webpack configuration to allow transpiling of @hsds modules (HSDS 4) for testing purposes, also we update react to the minimum version that permits the automatic runtime and make that change on babel too.

mariani-helpscout commented on src/components/SimpleModal/SimpleModal.jsx:
```diff
@@ -156,6 +159,7 @@ function SimpleModal({
     >
       <SimpleModalUI
         aria-modal="true"
+        aria-hidden={!isNil(rest.ariaHidden) ? rest.ariaHidden : !show}
```
You can use the [nullish coalescing operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) instead of bringing in a lodash function. I believe the same outcome can be accomplished with
```
aria-hidden={rest.ariaHidden ?? !show}
```

tinkertrain replied:
I'm aware and sometimes used to use something different to do the same (`something != null`). This is very explicit and what we use throughout the library. Maybe we can eventually replace all uses of lodash.isNil with that, thanks for the suggestion!

mariani-helpscout reviewed (commented):
One question and one non-required comment

tinkertrain commented on src/components/Input/__tests__/Input.test.js:
```diff
@@ -44,7 +41,7 @@ describe('ClassName', () => {
 describe('Input', () => {
   test('Can generate an input component', () => {
-    const wrapper = render(<Input />)
+    const wrapper = enzymeRender(<Input />)
```
The tests were written with Enzyme, I just renamed enzyme's `render` to `enzymeRender`, for clarity and to leave the name `render` for the testing library one, which I used for the new tests (previously renamed as `renderComponent` in this suite)

knicklabs reviewed (approved):
Thanks for tackling this @tinkertrain! That Safari screen reader issue is such a pain. I think you found a good solution though. And although it's a tad unfortunate that there's some manual bookkeeping involved when using the SimpleModal component, hopefully it's not necessary most of the time because hopefully nested modals are not a ux pattern we employ often :)

jakubjanczyk commented on src/components/Input/Input.jsx:
```diff
@@ -484,26 +493,19 @@ export class Input extends React.PureComponent {
-            </CharValidatorText>
-          </Badge>
-        </Animate>
+    // console.log('🚀 ~ getCharValidatorMarkup ~ isFocused', isFocused)
```
console.log left from testing I guess :)

jakubjanczyk reviewed (commented):
Code looks good, and most fixes as well, but for the input char validator badge issue, there is one case where it still happens:
![chrome-capture-2022-5-21](https://user-images.githubusercontent.com/1765264/174751674-1a9379aa-607b-4f20-ad23-f82f42cd376f.gif)

Steps to reproduce:
- type until limit reached
- select all the text
- replace old text with new by typing some new character (without removing the old one first)

tinkertrain commented:
@knicklabs I know... I have no idea why, there's nothing "special" about the SimpleModal, the focus _is_ transferred to the 2nd modal even on safari, just not with VoiceOver for some reason 😕 I do hope we don't use nested modals often, they don't feel like a great UX imo.

@jakubjanczyk Good find! I simplified the logic, I removed a check which honestly don't know what it was for and things now work in all circumstances (I think), added an extra test for good measure.

jakubjanczyk reviewed (approved):
Thanks, looks good :)
